arch/eventloop_posix_udp.c

```diff
@@ -36,6 +44,524 @@ typedef struct {
+    * */
+    UA_StatusCode res = UA_STATUSCODE_GOOD;
+    res |= UDP_setNonBlocking(socket);
+    res |= UDP_setNoSigPipe(socket);
+    if(res != UA_STATUSCODE_GOOD) {
+        UA_LOG_SOCKET_ERRNO_WRAP(
+            UA_LOG_WARNING(logger, UA_LOGCATEGORY_NETWORK,
+                           "UDP\t| Could not set socket options: %s", errno_str));
+        UA_close(socket);
+        return res;
+    }
+
+    return UA_STATUSCODE_GOOD;
+}
+static UA_StatusCode
+configureMulticastInterfaceIPv4(UA_FD socket,
+                                socklen_t addrlen, struct sockaddr_in *addr,
+                                size_t paramsSize, const UA_KeyValuePair *params, const UA_Logger *logger){
+    struct in_addr address = addr->sin_addr;
+    UA_Byte *addressVal = (UA_Byte *) &addr->sin_addr;
+
+    if(isIPv4MulticastAddress(addressVal)) {
+
+        IpMulticastRequest ipMulticastRequest;
+        ipMulticastRequest.ipv4.imr_multiaddr = address;
+        /* default outgoing interface ANY */
+        ipMulticastRequest.ipv4.imr_interface.s_addr = UA_htonl(INADDR_ANY);
+
+        UA_String netif = UA_STRING_NULL;
+        int foundInterface = getNetworkInterfaceFromParams(paramsSize, params, &netif, logger);
+        if(foundInterface < 0) {
+            UA_LOG_ERROR(logger, UA_LOGCATEGORY_NETWORK,
+                         "UDP\t| Opening a connection failed");
+            return UA_STATUSCODE_BADINTERNALERROR;
+        }
+
+        if(netif.length > 0) {
+            UA_STACKARRAY(char, interfaceAsChar, sizeof(char) * (netif).length + 1);
+            memcpy(interfaceAsChar, netif.data, netif.length);
+            interfaceAsChar[netif.length] = 0;
+
+            if(UA_inet_pton(AF_INET, interfaceAsChar, &ipMulticastRequest.ipv4.imr_interface) <= 0) {
+                UA_LOG_ERROR(logger, UA_LOGCATEGORY_SERVER,
+                             "PubSub Connection creation problem. "
+                             "Interface configuration preparation failed.");
+                return UA_STATUSCODE_BADINTERNALERROR;
+            }
+        }
+        setsockopt(socket, IPPROTO_IP, IP_MULTICAST_IF,
+                   &ipMulticastRequest.ipv4.imr_interface, sizeof(struct in_addr));
+    }
+    return UA_STATUSCODE_GOOD;
+}
+
+static UA_StatusCode
+addMulticastInterfaceIPv4(UA_FD socket,
+                          size_t paramsSize, const UA_KeyValuePair *params,
+                          socklen_t addrlen, struct sockaddr_in *addr,
+                          const UA_Logger *logger){
+    struct in_addr address = addr->sin_addr;
+    UA_Byte *addressVal = (UA_Byte *) &addr->sin_addr;
+
+    if(isIPv4MulticastAddress(addressVal)) {
+
+        IpMulticastRequest ipMulticastRequest;
+        ipMulticastRequest.ipv4.imr_multiaddr = address;
+        /* default outgoing interface ANY */
+        ipMulticastRequest.ipv4.imr_interface.s_addr = UA_htonl(INADDR_ANY);
+
+        UA_String netif = UA_STRING_NULL;
```

Comment:
You call getNetworkInterfaceParams two times separately.
That looks unnecessary and can be simplified.
Eg by merging the configuration and register steps.